### PR TITLE
Port Data.List.unzip from Idris1

### PR DESCRIPTION
**Pull request by ska80**

### CODE REVIEW

File: libs/base/Data/List.idr [hunk 1]
```diff
@@ -358,6 +358,20 @@ export
 zip3 : List a -> List b -> List c -> List (a, b, c)
 zip3 = zipWith3 \x, y, z => (x, y, z)
 
+||| Split a list of pairs into two lists
+export
+unzip : List (a, b) -> (List a, List b)
```
Comment (gallais) [inline libs/base/Data/List.idr:363]: Can we have `unzipWith : (a -> (b, c)) -> List a -> (List b, List c)` too? :pleading_face: 
I am always annoyed that we have `zipWith`, `zip`, `unzip` but no `unzipWith`.

Comment (stefan-hoeck) [reply libs/base/Data/List.idr:363]: Also, if we have `zip3`, we should probably also have `unzip3` and `unzipWith3`.

Comment (ska80) [reply libs/base/Data/List.idr:363]: > Can we have `unzipWith : (a -> (b, c)) -> List a -> (List b, List c)` too? 🥺
> I am always annoyed that we have `zipWith`, `zip`, `unzip` but no `unzipWith`.

Good idea.  I will add them.

Comment (stefan-hoeck) [reply libs/base/Data/List.idr:363]: Just one more note: If we add `unzipWith`, we should have this also for `Vect`, `Stream`, and `List1` (probably also for `Maybe` and `Either e`) to be consistent across different (un)zippable data types. I wonder, whether we actually should have an interface for this (like MonadZip in Haskell). Probably this should not all go in this PR, though?

Comment (ska80) [reply libs/base/Data/List.idr:370]: Added `zipWith` and `zipWith3`

File: libs/base/Data/List.idr [hunk 2]
```diff
@@ -358,19 +358,31 @@ export
 zip3 : List a -> List b -> List c -> List (a, b, c)
 zip3 = zipWith3 \x, y, z => (x, y, z)
 
+||| Split a list by applying a function into two lists
+export
+unzipWith : (a -> (b, c)) -> List a -> (List b, List c)
+unzipWith f [] = ([], [])
+unzipWith f (x :: xs) with (unzipWith f xs)
+  unzipWith f (x :: xs) | (lefts, rights) = (fst (f x) :: lefts, snd (f x) :: rights)
```
Comment (gallais) [inline libs/base/Data/List.idr:366]: I don't think we need to use a `with` (that means generating an auxiliary definition)
and we should probably avoid duplicating the `f x` call. So I'd write something like
this instead (and something similar for `unzipWith3`):

```suggestion
unzipWith f (x :: xs)
  = let (b, c) = f x
        (bs, cs) = unzipWith f xs
    in (b :: bs, c :: cs)
```

**Pull request merged**